src/ir/effects.h

```diff
@@ -199,14 +207,21 @@ struct EffectAnalyzer
         return true;
       }
     }
-    // we are ok to reorder implicit traps, but not conditionalize them
+    // We are ok to reorder implicit traps, but not conditionalize them.
     if ((implicitTrap && other.transfersControlFlow()) ||
         (other.implicitTrap && transfersControlFlow())) {
       return true;
     }
-    // we can't reorder an implicit trap in a way that alters global state
-    if ((implicitTrap && other.hasGlobalSideEffects()) ||
-        (other.implicitTrap && hasGlobalSideEffects())) {
+    // Note that the above includes disallowing the reordering of a trap with an
+    // exception (as an exception can transfer control flow inside the current
+    // function, so transfersControlFlow would be true) - while we allow the
+    // reordering of traps with each other, we do not reorder exceptions with
+    // anything.
+    assert(!((implicitTrap && other.throws) || (throws && other.implicitTrap)));
```

Comment:
```suggestion
    assert(!(implicitTrap && other.throws) && !(throws && other.implicitTrap));
```
Nit: I personally find this expansion easier to read..